### PR TITLE
Make sstable reader fail on unknown column names

### LINKED ISSUE #2598: Sstable reader should fail on unknown column names
Currently it drops them, which is not safe. We risk data loss due to a bug. Refs #2597. Dropped columns would still be recognized, because they're recorded in schema.

### PR DESCRIPTION
**Pull request by haaawk**
Fixes #2598.

Tests: unit(dev)

### CODE REVIEW

**Comment by haaawk** [comment]
@avikivity @tgrabiec @bhalevy please have a look

File: sstables/mp_row_consumer.hh [hunk 1]
```diff
@@ -509,6 +509,11 @@ public:
     //}
     proceed do_consume_cell(bytes_view col_name, int64_t timestamp, int64_t ttl, int64_t expiration, CreateCell&& create_cell) {
         struct column col(*_schema, col_name, timestamp);
+        auto cell_str = sstring(to_sstring_view(col.cell));
+        auto dc = _schema->dropped_columns().find(cell_str);
+        if (!col.cdef && col.cell.size() > 0 && (dc == _schema->dropped_columns().end() || dc->second.timestamp < timestamp)) {
```
Comment (bhalevy) [inline sstables/mp_row_consumer.hh]: How about evaluating dc only `if (!col.cdef && col.cell.size() > 0)`?

File: sstables/mp_row_consumer.hh [hunk 2]
```diff
@@ -576,6 +581,11 @@ public:
     virtual proceed consume_deleted_cell(bytes_view col_name, sstables::deletion_time deltime) override {
         auto timestamp = deltime.marked_for_delete_at;
         struct column col(*_schema, col_name, timestamp);
+        auto cell_str = sstring(to_sstring_view(col.cell));
+        auto dc = _schema->dropped_columns().find(cell_str);
+        if (!col.cdef && col.cell.size() > 0 && (dc == _schema->dropped_columns().end() || dc->second.timestamp < timestamp)) {
```
Comment (bhalevy) [inline sstables/mp_row_consumer.hh]: ditto

File: sstables/mp_row_consumer.hh (hunk 1 above)
Comment (haaawk) [reply sstables/mp_row_consumer.hh]: I optimised the condition and pulled the code to a single helper function.

File: sstables/mp_row_consumer.hh (hunk 2 above)
Comment (haaawk) [reply sstables/mp_row_consumer.hh]: done

**Comment by avikivity** [comment]
Why should we fail on unknown column names? The cover letter should explain this.

IMO this is wrong:
1. create table on cluster A
2. write some data
3. delete column from table
4. recreate schema (after column deletion) on cluster B
5. import sstable from cluster A
6. import fails

**Comment by haaawk** [comment]
It's a fair point @avikivity but you can always use initial schema for B population and then remove the column again.

@tgrabiec: you were the author of the issue so maybe you could comment :)

Maybe printing warning to logs instead of an exception would be a better solution?

**Comment by bhalevy** [comment]
> Maybe printing warning to logs instead of an exception would be a better solution?

If we consider this, I think it merits a configuration option to allow the user to decide about the semantics.
If possible, we should set the default to the safe behavior (throwing an exception) at least for new installations (in `conf/scylla.yaml`) and to the current behavior (ignoring / warning) on upgrades (as the hard coded default if missing from `scylla.conf`) for backward compatibility.

**Comment by avikivity** [comment]
> It's a fair point @avikivity but you can always use initial schema for B population and then remove the column again.

You could, if you anticipated the problem. Most people will not.

> 
> Maybe printing warning to logs instead of an exception would be a better solution?

We can do that for mc sstables where the column names are known in advance. We can ignore ka/la, they are obsolete.

**Comment by haaawk** [comment]
> > Maybe printing warning to logs instead of an exception would be a better solution?
> 
> We can do that for mc sstables where the column names are known in advance. We can ignore ka/la, they are obsolete.

At the moment we throw exception in mc already.

**Comment by avikivity** [comment]
Then we should change it to log, to prevent the scenario from above failing.

**Comment by tgrabiec** [comment]
@avikivity In general your scenario won't work, because the information about dropped columns is needed to properly interpret the sstables. If some column was dropped in cluster A and then created again, you need the information about dropped columns to filter out writes in sstables.

If we import data by copying sstables, we should copy the schema sstables as well. Or have a tool for exporting and importing schema properly.

Dropping writes to unknown tables seems very unsafe. A bug could cause compaction to cause data loss. Refs https://github.com/scylladb/scylla/issues/2597.

**Comment by avikivity** [comment]
@glommer you're a copy-sstables-to-migrate kind of guy, what's your take?

**Comment by glommer** [comment]
It's hard. From the user level, we can't really match a schema to an SSTable.

Let's say you delete a column. The existing SSTables are not aware of that, as far as I understand. The user will import the schema through a `DESC` command, which will not contain the updated schema only.

**Comment by tgrabiec** [comment]
@glommer What do you suggest as a way forward? 

Do we agree that importing data by copying the sstables and at the same time importing schema by using `DESC` is in general incorrect?

I think we either need to import data using sstableloader or have a way to migrate full schema definition to another cluster. There already is a way if you want to migrate all tables (copy the schema tables). We could have a command for duping a slice of schema tables for a particular table.

**Comment by glommer** [comment]
I don't agree that using DESC and importing data is incorrect - there is no other way around it and we've done it many times.

Even for backups: Let's say you are taking backups from the cluster and the cluster dies. Our documented procedure is to restore the schema and then load the SSTables.

How is that supposed to work ?

**Comment by haaawk** [comment]
@glommer The is a way around poor work that DESC does. It's enough to read full schema description from system tables with CQL and then write it to the new/restored cluster. We can have a simple tool for that. This way all information about deleted columns would be preserved. It would work with backups too.

**Comment by tgrabiec** [comment]
On Wed, Jul 10, 2019 at 11:38 PM Glauber Costa <notifications@github.com>
wrote:

> I don't agree that using DESC and importing data is incorrect - there is
> no other way around it and we've done it many times.
>

The fact that we've done it many times doesn't make it correct. I gave an
example of when it breaks, and proposed a solution.


> Even for backups: Let's say you are taking backups from the cluster and
> the cluster dies. Our documented procedure is to restore the schema and
> then load the SSTables.
>
> How is that supposed to work ?
>
This procedure is wrong, for the same reasons. Schema should be backed up
by snapshotting schema tables.

>

**Comment by avikivity** [comment]
We could have a version of DESCRIBE that generates the schema with deleted/renamed columns documented, and CQL extensions for deleted/renamed columns:

```
DESCRIBE my_ks.my_tab FULL

CREATE TABLE my_ks.my_tab (
    id INT PRIMARY KEY
    c1 INT,
    c2 INT DELETED AT 01-APR-2019,
    c3 INT RENAMED FROM c4
)
```

That would require migrations to be using our cqlsh against Cassandra, at least when there have been column deletions. Implementing the cql extensions should not be hard.

**Comment by haaawk** [comment]
This won't work with Cassandra clusters because Cassandra won't understand the extension. Did you think about cqlsh extension instead of CQL extension?.

**Comment by avikivity** [comment]
DESCRIBE is implemented in cqlsh so no Cassandra changes are needed, just cqlsh.

The cql changes are needed to import that schema.

**Comment by avikivity** [comment]
to be clear, both cqlsh and cql extensions are needed, the former for DESCRIBE, the latter for CREATE TABLE.

**Comment by haaawk** [comment]
Wouldn't it be easier to just provide a tool that reads schema description using CQL and then uploads it to a different cluster using CQL as well? All by simply modifying system tables.

**Comment by avikivity** [comment]
The tool to read the schema tables exists, it is called cqlsh. It also has the ability to create a schema on a different cluster by executing those commands. 

Modifying system tables doesn't create new tables, because the cluster doesn't perform the side effects. You also have to make the changes on every node.

**Comment by haaawk** [comment]
So it seems that this change is appropriate and we should additionally modify cqlsh and CQL to make it possible for proper schema export/import, right?

**Comment by avikivity** [comment]
I'm not sure we should touch la/ka, since we won't create new sstables with those formats.

**Comment by glommer** [comment]
I am not saying that we need to keep doing things a certain way just because we're already doing it, but I am saying that we need a user-friendly way of doing it.

`DESC SCHEMA` is such a way, system tables is not. For starters, it doesn't work cross versions (while importing SSTables cross versions is fine as long as token ownership is the same), and we have a variety of users that tried to restore system table schema on slack and made an almost unrecoverable mess out of their clusters.

Also once the procedure is out in the open, people keep using it.

I like Avi's idea of enriching `DESC schema`.

**Comment by tgrabiec** [comment]
On Thu, Jul 11, 2019 at 3:50 PM Glauber Costa <notifications@github.com>
wrote:

> I am not saying that we need to keep doing things a certain way just
> because we're already doing it, but I am saying that we need a
> user-friendly way of doing it.
>
> DESC SCHEMA is such a way, system tables is not. For starters, it doesn't
> work cross versions (while importing SSTables cross versions is fine as
> long as token ownership is the same),
>
Can you say more on why doesn't it work? Importing schema tales from older
to a newer version should work, like during regular upgrade.

Importing to a lower version may not, but so is true for data sstables, and
for the new "DESC schema" syntax.

and we have a variety of users that tried to restore system table schema on
> slack and made an almost unrecoverable mess out of their clusters.
>
> What was the cause?

> Also once the procedure is out in the open, people keep using it.
>
Are you saying this as an argument against throwing an exception, because
people using the current procedure would get an exception in case they have
data in sstables which was dropped? It doesn't seem like a big problem,
because if we have a new procedure in place, we can tell them to use the
new procedure, and their problem will be solved.


> I like Avi's idea of enriching DESC schema.
>

**Comment by glommer** [comment]
I have no opinion on whether throwing an exception is the right way to handle this. I am focusing my argument around whether or not it is reasonable to forego `DESC schema`, which I think it's not.

It's a user friendly process that works for both Cassandra and Scylla. Would importing system tables from Cassandra work in Scylla ?

**Comment by bhalevy** [comment]
@haaawk said:
> So it seems that this change is appropriate and we should additionally modify cqlsh and CQL to make it possible for proper schema export/import, right?

@avikivity replied:
> I'm not sure we should touch la/ka, since we won't create new sstables with those formats.

@haaawk change LGTM modulo exception controversy.  IMO it's safer to throw an exception in this case. The suggested follow up project at cqlsh/CQL level to allow migrating dropped columns between clusters will prevent these errors, and until then I assume that a user could resort to sstableloader or to modifying the schema on the destination cluster to create and drop the missing columns as needed.

**Pull request closed without merging**